pkg/ensign/api/v1beta1/deduplication.go

```diff
@@ -0,0 +1,194 @@
+	if _, err = hash.Write(event.Data); err != nil {
+		return nil, fmt.Errorf("could not write event data to hash: %w", err)
+	}
+
+	// Sort the keys in the metadata to write them in lexicographic order
+	keys := make([]string, 0, len(event.Metadata))
+	for key := range event.Metadata {
+		keys = append(keys, key)
+	}
+	slices.Sort(keys)
+
+	// Write the metadata to the to the hash
+	for _, key := range keys {
+		val := event.Metadata[key]
+		if _, err = hash.Write([]byte(key + val)); err != nil {
+			return nil, fmt.Errorf("could not write metadata key value pair to hash: %w", err)
+		}
+	}
+
+	// Write the mimetype to the hash
+	if err = binary.Write(hash, binary.LittleEndian, event.Mimetype); err != nil {
+		return nil, fmt.Errorf("could not hash mimetype: %w", err)
+	}
+
+	// Write the type to the hash
+	var etype *Type
+	if event.Type != nil && !event.Type.IsZero() {
+		etype = event.Type
+	} else {
+		etype = UnspecifiedType
+	}
+
+	if _, err = hash.Write([]byte(etype.Name)); err != nil {
+		return nil, fmt.Errorf("could not write type name to hash: %w", err)
+	}
+
+	if err = binary.Write(hash, binary.LittleEndian, etype.MajorVersion); err != nil {
+		return nil, fmt.Errorf("could not hash type major version: %w", err)
+	}
+
+	if err = binary.Write(hash, binary.LittleEndian, etype.MinorVersion); err != nil {
+		return nil, fmt.Errorf("could not hash type minor version: %w", err)
+	}
+
+	if err = binary.Write(hash, binary.LittleEndian, etype.PatchVersion); err != nil {
+		return nil, fmt.Errorf("could not hash type patch version: %w", err)
+	}
+
+	return hash.Sum(nil), nil
+}
+
+// Datagram hashing is used to detect duplicates in data only, ignoring metadata,
+// mimetype, and type as in strict hashing. This method returns a murmur3 hash of the
+// data field of the event only.
+func (w *EventWrapper) HashDatagram() (_ []byte, err error) {
+	var event *Event
+	if event, err = w.Unwrap(); err != nil {
+		return nil, err
+	}
+
+	hash := murmur3.New128()
+	if _, err = hash.Write(event.Data); err != nil {
+		return nil, err
+	}
+
+	return hash.Sum(nil), nil
+}
+
+// Key grouped hashing returns the murmur3 hash of the data of the event prefixed with
+// the metadata values of the the specified keys. E.g. if the data is foobar and the
+// hash is grouped by the key month - then for two events with month jan and month feb
+// will have different hashes: murmur3(janfoobar) and murmur3(febfoobar).
+//
+// NOTE: this method does not take into account mimetype or type but in the future we
+// may have "reserved keys" to factor in these elements to the hash.
+func (w *EventWrapper) HashKeyGrouped(keys []string) (_ []byte, err error) {
+	if len(keys) == 0 {
+		return nil, ErrNoKeys
+	}
+
+	var event *Event
+	if event, err = w.Unwrap(); err != nil {
+		return nil, err
+	}
+
+	hash := murmur3.New128()
+	for _, key := range keys {
+		val := event.Metadata[key]
+		if _, err = hash.Write([]byte(key + val)); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
the order of the keys specified to this function will determine the order of the key value pairs in the hash. I'm also concatenating the key + value in the hash but I'm not sure this is necessary. E.g. what if the event has none of the keys specified by the policy, we'd just be writing the key names to the event. This is a good thing if the hash should not be stable wrt policy changes (e.g. adding or removing keys) … but I'm not sure it's adding anything to the code.

Reply:
I see your point, I wonder if we could make that configurable? So it can either be unique by keys or also by key+value?

Reply:
I think concatenating the key + value in the hash works for bookkeeping purposes, we know which keys were used for hashing.  This is also helpful for the reason you pointed out, that the hash should not be stable wrt policy changes.

Reply:
Ok - thank you for your input!